### PR TITLE
[DPDV-5997] feat: allow query all or selected accounts

### PR DESCRIPTION
**Pull request by munna-shaik-s1**
Jira Link: <https://sentinelone.atlassian.net/browse/DPDV-5997>

# 🥅 Goal

Allow to query all assigned accounts or specify list of accounts to query

# 🛠️ Solution

- add support for AuthN tokens
- allow query all assigned accounts via tenant parameter
- allow query selected accounts via accountIds parameter

# 🏫 Testing

How the changes were tested

### CODE REVIEW

File: README.md
Comment (zdaratom-s1) [inline README.md:74]: Are those actual accountIds? If so please use dummy values rather eg. `123456789, 987654321`

File: globalConfig.json
Comment (zdaratom-s1) [inline globalConfig.json:137]: Would be great to update this according to read me, eg. optional on case of ...

File: globalConfig.json
Comment (zdaratom-s1) [inline globalConfig.json:74]: Please describe behaviour when its Blank

File: TA_dataset/README/ta_dataset_settings.conf.spec
Comment (zdaratom-s1) [inline TA_dataset/README/ta_dataset_settings.conf.spec:5]: Why is the field name so cryptic? would it be possible to call it eg. `authn_token_part_one` ?

File: TA_dataset/bin/dataset_api.py
Comment (zdaratom-s1) [inline TA_dataset/bin/dataset_api.py:65]: Duplicate code, can we externalise this to separate method?

File: TA_dataset/bin/dataset_api.py
Comment (zdaratom-s1) [inline TA_dataset/bin/dataset_api.py:135]: Why dont we unify the naming here? `tenant_details vs acc_conf` Having 2 different name for same variable make it bit more unreadable

File: TA_dataset/bin/dataset_api.py
Comment (zdaratom-s1) [inline TA_dataset/bin/dataset_api.py:266]: what is the purpose of this change?

File: TA_dataset/bin/dataset_common.py
Comment (zdaratom-s1) [inline TA_dataset/bin/dataset_common.py:184]:
```suggestion
                        logger.info("The AuthN api token was not available")
```

File: TA_dataset/bin/dataset_common.py
Comment (zdaratom-s1) [inline TA_dataset/bin/dataset_common.py:170]:
```suggestion
                    authn_token = ""
```

File: TA_dataset/bin/dataset_common.py
Comment (zdaratom-s1) [inline TA_dataset/bin/dataset_common.py:192]: Is not account_ids should be filled only if  tenant = false?

File: TA_dataset/bin/dataset_common.py
Comment (zdaratom-s1) [inline TA_dataset/bin/dataset_common.py:207]: code duplication

File: TA_dataset/bin/dataset_common.py
Comment (zdaratom-s1) [inline TA_dataset/bin/dataset_common.py:247]: code duplication

File: TA_dataset/bin/dataset_common.py
Comment (zdaratom-s1) [inline TA_dataset/bin/dataset_common.py:311]: its not `rw` anymore, lets rename this parameter

File: TA_dataset/bin/dataset_common.py
Comment (zdaratom-s1) [inline TA_dataset/bin/dataset_common.py:328]: consider chaning value `read` and `write` to apitoken_read/write so its more explicit

File: TA_dataset/bin/dataset_common.py
Comment (zdaratom-s1) [inline TA_dataset/bin/dataset_common.py:311]:
```suggestion
def get_token(self, account, token_type, logger, config_key=None):
```

File: TA_dataset/bin/dataset_common.py
Comment (zdaratom-s1) [inline TA_dataset/bin/dataset_common.py:296]: imho this statement is redundant.

File: TA_dataset/bin/dataset_common.py
Comment (zdaratom-s1) [inline TA_dataset/bin/dataset_common.py:301]: Do we validate trailing comma? eg. `accountId1,accountId2,` ideally we ignore empty part after trailing comma or rise an validation exception

File: TA_dataset/bin/dataset_query.py
Comment (zdaratom-s1) [inline TA_dataset/bin/dataset_query.py:147]: would it be possible to extract this code duplication to separate method?

File: TA_dataset/bin/dataset_search_command.py
Comment (zdaratom-s1) [inline TA_dataset/bin/dataset_search_command.py:445]: another code duplication

File: globalConfig.json
Comment (zdaratom-s1) [inline globalConfig.json:2]: would it be possible to avoid reformating of the file? it difficult to review these changes

**Review by zdaratom-s1 — COMMENTED** [review]
Please update PR description and title so its clear what is your goal

File: globalConfig.json
Comment (munna-shaik-s1) [reply globalConfig.json:137]: If you have any thoughts please provide the help string here. will update it with the same

File: TA_dataset/README/ta_dataset_settings.conf.spec
Comment (munna-shaik-s1) [reply TA_dataset/README/ta_dataset_settings.conf.spec:5]: because there is limitations on the splunk encrypted keys. in order to achieve that, used this type of names

File: TA_dataset/bin/dataset_api.py
Comment (munna-shaik-s1) [reply TA_dataset/bin/dataset_api.py:259]: It was done automatically after running the pre-commit hook, as a part of linting fix

File: TA_dataset/bin/dataset_common.py
Comment (zdaratom-s1) [reply TA_dataset/bin/dataset_common.py]: My point was to remove the return statement now whole section

File: TA_dataset/bin/dataset_common.py
Comment (munna-shaik-s1) [reply TA_dataset/bin/dataset_common.py]: we do supporting the read and write access SDL token as well, that the reason we don't changed the existing ones

File: TA_dataset/bin/dataset_query.py
Comment (munna-shaik-s1) [reply TA_dataset/bin/dataset_query.py:147]: this is called with in the dataset_query file only, even if we do separate method this method has only one calling method in this file

File: TA_dataset/bin/dataset_search_command.py
Comment (munna-shaik-s1) [reply TA_dataset/bin/dataset_search_command.py:445]: same here

File: TA_dataset/bin/dataset_common.py
Comment (munna-shaik-s1) [reply TA_dataset/bin/dataset_common.py:241]: already handled

File: TA_dataset/README/ta_dataset_settings.conf.spec
Comment (zdaratom-s1) [reply TA_dataset/README/ta_dataset_settings.conf.spec:5]: what is the limitation?

File: globalConfig.json
Comment (munna-shaik-s1) [reply globalConfig.json:2]: i made only in the account configuration section only. not sure why its showing these many lines code change.

File: TA_dataset/bin/dataset_common.py
Comment (zdaratom-s1) [reply TA_dataset/bin/dataset_common.py]: Do you mean this value is stored in Splunk config? So in order to not break backwards compatibility its not possible?

File: TA_dataset/bin/dataset_query.py
Comment (zdaratom-s1) [reply TA_dataset/bin/dataset_query.py:147]: There are 3 similar blocks
- https://github.com/scalyr/dataset-addon-for-splunk/pull/127/files#diff-d18f827fd8bcb2d6d19de613abcf89f511869f92e201edc5521bb16705880e7cR123
- https://github.com/scalyr/dataset-addon-for-splunk/pull/127/files#diff-82c2c47a7187c5d5ae31339c5b9a97eb6760d164afac8c973e29d5ed4c95f200R116
- https://github.com/scalyr/dataset-addon-for-splunk/pull/127/files#diff-61e399a15e16c531564f016785921ec4e678e17ec74380722f9abf2788de528fR150

File: TA_dataset/bin/dataset_search_command.py
Comment (zdaratom-s1) [reply TA_dataset/bin/dataset_search_command.py:445]: This duplication is across multiple files

File: globalConfig.json
Comment (zdaratom-s1) [reply globalConfig.json:137]: Do I understand https://github.com/scalyr/dataset-addon-for-splunk/pull/127/files#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5R80 correctly that it is optional in case of AuthN is provided, right?

Comment (munna-shaik-s1) [reply globalConfig.json]: Done

File: TA_dataset/README/ta_dataset_settings.conf.spec
Comment (munna-shaik-s1) [reply TA_dataset/README/ta_dataset_settings.conf.spec:5]: all encrypted values are saved in single instance of Storage Manager which does not allow data <256 chars

File: TA_dataset/bin/dataset_common.py
Comment (munna-shaik-s1) [reply TA_dataset/bin/dataset_common.py]: correct in-order to achieve the backward compatibility

File: TA_dataset/bin/dataset_query.py
Comment (munna-shaik-s1) [reply TA_dataset/bin/dataset_query.py]: Done

File: TA_dataset/README/ta_dataset_settings.conf.spec
Comment (zdaratom-s1) [reply TA_dataset/README/ta_dataset_settings.conf.spec:5]: Yes I understand that from description, my point is not the value of this variable but variable name, which I found bit difficult to decrypt `an_fir_part `, I suggest eg. `authn_token_part_one `

File: globalConfig.json
Comment (zdaratom-s1) [inline globalConfig.json:137]: I suppose same should go to `dataset_log_write_access_key` right?

I would prefer eg. `Required (if no AuthN token provided) to enable inputs and SPL comand. Include trailing hyphens if applicable.`

Comment (munna-shaik-s1) [reply globalConfig.json:137]: No, write key is different, this only applicable to read ones only

File: TA_dataset/bin/dataset_common.py
Comment (munna-shaik-s1) [reply TA_dataset/bin/dataset_common.py]: this is just for identifying the type of the key, read/write

File: globalConfig.json
Comment (zdaratom-s1) [inline globalConfig.json:137]:
```suggestion
                            "help": "Required (if no AuthN token provided) to enable inputs and SPL command. Include trailing hyphens if applicable.",
```

File: TA_dataset/bin/dataset_api.py
Comment (zdaratom-s1) [inline TA_dataset/bin/dataset_api.py:47]: This method seems to be duplicate with get_tenant_related_payload()

File: TA_dataset/bin/dataset_common.py
Comment (zdaratom-s1) [reply TA_dataset/bin/dataset_common.py:251]: where? `.strip()` handles whitespaces not comma

File: TA_dataset/bin/dataset_common.py
Comment (zdaratom-s1) [reply TA_dataset/bin/dataset_common.py]: I am confused this is not needed anymore as you have deleted it?

File: TA_dataset/bin/dataset_common.py
Comment (munna-shaik-s1) [reply TA_dataset/bin/dataset_common.py:251]: Splunk provides the config validations it uses the regex as the validators. For account Ids used this regex `^\\d+(,\\d+)*$` as a validator. this will check the comma trialing exception, if any trialing comma occurs it wont allow the user to save the configuration.

File: TA_dataset/bin/dataset_common.py
Comment (zdaratom-s1) [reply TA_dataset/bin/dataset_common.py]: Yes and that was my point that i would be better to change this identifier in order to improve readability

File: globalConfig.json
Comment (zdaratom-s1) [reply globalConfig.json:2]: I have checked the changes despite different spacing

File: TA_dataset/lib/dataset_query_api_client/api/default/post_queries.py
Comment (zdaratom-s1) [inline TA_dataset/lib/dataset_query_api_client/api/default/post_queries.py:24]: imho no tenant_details is needed as PostQueriesLaunchQueryRequestBody already contains field additional_properties which should be sufficient

File: TA_dataset/bin/dataset_api.py
Comment (zdaratom-s1) [inline TA_dataset/bin/dataset_api.py:112]: Lets pass response from get_tenant_related_payload(acc_conf) to `PostQueriesLaunchQueryRequestBody.additional_properties` so we can avoid introducing additional attributes

**Review by zdaratom-s1 — COMMENTED** [review]
One more small improvement

File: TA_dataset/bin/dataset_api.py
Comment (munna-shaik-s1) [reply TA_dataset/bin/dataset_api.py]: done

File: TA_dataset/bin/dataset_common.py
Comment (munna-shaik-s1) [reply TA_dataset/bin/dataset_common.py]: removed that if tenant condition. we need this method to fetch the account ids. Hope this make sense

**Comment by zdaratom-s1** [comment]
blocked by https://sentinelone.atlassian.net/browse/DPDV-7764

**Comment by zdaratom-s1** [comment]
Please rebase, test are fixed in main branch already

**Review by zdaratom-s1 — APPROVED** [review]

**Pull request merged**
